lib/iolog/log_writer.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iolog
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"time"
+)
+
+const (
+	// MaxEntrySizeBytes is the maximum allowable size of a log entry
+	// before it will be broken into multiple entries
+	MaxEntrySizeBytes = 16 * 1024
+)
+
+// LogWriter tags log entries with descriptive metadata
+type LogWriter struct {
+	w io.Writer
+}
+
+// NewLogWriter wraps an io.WriteCloser in a LogWriter
+func NewLogWriter(w io.Writer) *LogWriter {
+	return &LogWriter{w: w}
+}
+
+// Write scans the supplied buffer, breaking off indvidual entries
+// and writing them to the underlying Writer, flusing any leftover bytes
+func (lw *LogWriter) Write(p []byte) (n int, err error) {
+	start := 0
+	for {
+		i := 0
+		if i = bytes.IndexByte(p[start:], '\n') + 1; i == 0 {
+			break
+		}
+		end := start + i
+		entry := p[start:end]
+		// we have a complete entry, let's write it
+		n, err = lw.write(entry)
+		if err != nil {
+			return n, err
+		}
+		// advance starting index
+		start = end
+	}
+	// flush the buffer if needed
+	if start < len(p) {
+		lw.write(p[start:])
+	}
+	// inform the caller that all bytes were written successfully
+	return len(p), err
+}
+
+// write breaks a log entry into smaller entries if necessary, and
+// adds a header to each entry before writing to the underlying Writer
+func (lw *LogWriter) write(b []byte) (n int, err error) {
+	entries := [][]byte{}
+	// break the entry up into multiple entries if necessary
+	for len(b) > MaxEntrySizeBytes {
+		entries = append(entries, b[:MaxEntrySizeBytes])
+		b = b[MaxEntrySizeBytes:]
+	}
+	entries = append(entries, b)
+	for i := 0; i < len(entries); i++ {
+		entry := entries[i]
+		partial := 1
+		if i == len(entries)-1 {
+			// all entries are partial unless they are the last (or only) entry
+			partial = 0
+		} else {
+			// add a trailing newline to a partial entry
+			entry = append(entry, '\n')
+		}
+		// Each entry has a 10-byte header that describes the entry as follows:
+		// The first 8 bytes are the timestamp in int64 unix epoch format
+		// The first 14 bits of the final two bytes represent the size of the entry
+		// 0x2 represents the stream - 0 for stdout, 1 for stderr
+		// 0x1 represents whether or not the entry is a partial entry - 0 for complete, 1 for a partial entry.
+		header := make([]byte, 10)
```

Comment:
I think you should have an encoder and decoder for your LogEntry struct that has this code. Should make it cleaner and easier to test.